feat(ngmesh2obj): produce one file per mesh

william-silversmith opened the pull request:
Now produces a unifed mesh file instead of lots of mesh fragment files.
Also a major performance boost.

nkemnitz commented:
The fragment data does not contain vertex normals?

william-silversmith commented:
Nah, it seems neuroglancer computes them after download. Meshlab seems to
get along as well. Would you like them to be there?

On Jul 26, 2017 10:03 PM, "Nico Kemnitz" <notifications@github.com> wrote:

> The fragment data does not contain vertex normals?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/seung-lab/neuroglancer/pull/161#issuecomment-318237030>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACZoSfqtOv2MP8JsAbaBWAzIHe57P3y8ks5sR_AFgaJpZM4Okkaz>
> .
>

nkemnitz commented:
No, it's fine. If we kept the normals from marching cubes (before simplification), then they might be useful. But since we recalculate them from the already simplified meshes, they don't provide any additional information.

nkemnitz commented on python/neuroglancer/pipeline/scripts/ngmesh2obj.py:
```diff
@@ -48,57 +48,52 @@ def decode_downloaded_data(frag_datas):
+    objdata = []
+    
+    for vertex in fragment['vertices']:
+        objdata.append('v %s %s %s' % (vertex[0], vertex[1], vertex[2]))
```
Switching between 2 and 4 space indentation across the file

nkemnitz commented on python/neuroglancer/pipeline/scripts/ngmesh2obj.py:
```diff
@@ -23,7 +23,7 @@
 import gzip
```
`StringIO`, `gzip` and `neuroglancer.lib.mkdir` seem to be unused.

william-silversmith commented on python/neuroglancer/pipeline/scripts/ngmesh2obj.py:
```diff
@@ -48,57 +48,52 @@ def decode_downloaded_data(frag_datas):
 
 def decode_mesh_buffer(fragment):
-  """
-  Mesh array format is: 
```
See the top comment

william-silversmith commented:
I agree that the script should error for these conditions. When generating objs from a few tests, there seems to be usually one fragment that's screwed up somehow. Maybe there's a bug in MeshTask. I feel like until I debug it, notification that the obj is possibly missing a piece is enough so at least you can generate meshes to look at. 

The alternative is that I fix MeshTask and rerun meshing when we find screwed up meshes.

Here's an example error I'm seeing:

```
(venv)wms:~/neuroglancer/python/neuroglancer$ python pipeline/scripts/ngmesh2obj.py gs://neuroglancer/pinky40_v11/watershed_mst_smc_sem5_remap/ 27711747
downloading 27711747...
decoding...
Unable to process fragment mesh_mip_3_err_20/27711747:0. Violation: len vertex data != 12 * num vertices
        Array Length: 1919296123, Vertex Count: 4602
      
saving to obj...
done.
```

william-silversmith commented:
Found the bug. Sometimes the JSON manifest file was included in the downloaded fragment data. Obviously it's not a valid mesh. :P I added some handling code to exclude it if its accidentally included. I changed the warnings to errors now that the meshes are downloading and generating cleanly.

nkemnitz commented on python/neuroglancer/pipeline/scripts/ngmesh2obj.py:
```diff
@@ -19,86 +19,91 @@
+    faces = []
+    for i in xrange(0, len(face_data), 4):
+      vertex_number = struct.unpack("=I", face_data[i:i+4])[0]
+      if vertex_number > num_vertices:
```
should be `>=`

william-silversmith replied:
I thought they were numbered starting at 1?

william-silversmith replied:
Oh right, this is len(array) :P
